### PR TITLE
Vfb 40 sort by status

### PR DESCRIPTION
**Pull request by AmeliaIsCoding**
## What's changed
Sorting by last status in the parcel table results in rows ordered to be consistent with workflow - eg packed => ready to dispatch => delivered. Previously, sorting by last status had no effect except to revert any other sorts.

## Screenshots / Videos
| Before     | After      |
|------------|------------|
|imageimage|imageimage|

## Checklist
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
  - Will also be automatically checked on pull request
- [x] Make sure you've tested via `npm run test`
  - Will also be automatically checked on pull request

### CODE REVIEW

File: src/app/parcels/ParcelsPage.tsx [hunk 1]
```diff
@@ -56,6 +56,53 @@ const defaultShownHeaders: (keyof ParcelsTableRow)[] = [
...
+    "Collection Failed",
+    "Parcel Collected",
+    "Shipping Labels Downloaded",
+    "Out for Delivery",
+    "Delivered",
+    "Delivery Failed",
+    "Delivery Cancelled",
+    "Fulfilled with Trussell Trust",
+    "Request Deleted",
+];
```
[the remark this answers is not in the document]
Comment (AmeliaIsCoding) [reply src/app/parcels/ParcelsPage.tsx:78]: The statuses came from parcels/ActionBar/Statuses.tsx, but the order came from my heart...
Should I ask Lincoln to look over the order and see what he thinks?

File: src/app/parcels/ParcelsPage.tsx [hunk 2]
```diff
@@ -56,6 +56,53 @@ const defaultShownHeaders: (keyof ParcelsTableRow)[] = [
...
+    const indexB = statusNamesInWorkflowOrder.indexOf(statusB?.name ?? "-");
+    if (indexA > indexB) {
+        return -1;
+    } else if (indexB > indexA) {
+        return 1;
+    }
+    return 0;
+};
+
+const sortable: (keyof ParcelsTableRow | SortOptions<ParcelsTableRow, any>)[] = [
```
[the remark this answers is not in the document]
Comment (AmeliaIsCoding) [reply src/app/parcels/ParcelsPage.tsx:94]: ty!

File: src/app/parcels/ParcelsPage.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (rickaw) [reply src/app/parcels/ParcelsPage.tsx:78]: The order looks sensible to me. It's not ideal for us to have to maintain two separate lists of statuses though. Can we have just one list and import it from both places?

[the remark this answers is not in the document]
Comment (AmeliaIsCoding) [reply src/app/parcels/ParcelsPage.tsx]: Have done

File: src/app/parcels/ParcelsPage.tsx [hunk 3]
```diff
@@ -56,6 +57,32 @@ const defaultShownHeaders: (keyof ParcelsTableRow)[] = [
     "lastStatus",
 ];
 
+const sortStatusByWorkflowOrder = (
+    statusA: ParcelsTableRow["lastStatus"],
+    statusB: ParcelsTableRow["lastStatus"]
+): number => {
+    const indexA = statusNamesInWorkflowOrder.indexOf(statusA?.name ?? "-");
+    const indexB = statusNamesInWorkflowOrder.indexOf(statusB?.name ?? "-");
```
Comment (rickaw) [inline src/app/parcels/ParcelsPage.tsx:65]: I think the default value can be any string that's not in the array - `indexOf` would return -1 if the string is not found.

File: src/app/parcels/ParcelsPage.tsx [hunk 4]
```diff
@@ -56,6 +57,32 @@ const defaultShownHeaders: (keyof ParcelsTableRow)[] = [
...
+    if (indexA > indexB) {
+        return -1;
+    } else if (indexB > indexA) {
+        return 1;
+    }
+    return 0;
```
Comment (rickaw) [inline src/app/parcels/ParcelsPage.tsx:71]:
```suggestion
    const indexA = statusNamesInWorkflowOrder.indexOf(statusA?.name ?? "");
    const indexB = statusNamesInWorkflowOrder.indexOf(statusB?.name ?? "");
    
    if (indexA === indexB) return 0;
    return indexA < indexB ? 1 : -1
```

File: src/app/parcels/ParcelsPage.tsx (hunk 3 above)
Comment (AmeliaIsCoding) [reply src/app/parcels/ParcelsPage.tsx:65]: That makes sense thanks

File: src/app/parcels/ActionBar/Statuses.tsx [hunk 5]
```diff
@@ -10,27 +10,28 @@ import { ParcelsTableRow } from "@/app/parcels/getParcelsTableData";
 import StatusesBarModal from "@/app/parcels/ActionBar/StatusesModal";
 import { DatabaseError } from "@/app/errorClasses";
 
-const statuses = [
+export const statusNamesInWorkflowOrder = [
+    "-",
```
[the remark this answers is not in the document]
Comment (AmeliaIsCoding) [reply src/app/parcels/ActionBar/Statuses.tsx:14]: have removed, no longer need if we use "" as the default in the sort function

File: src/app/parcels/ActionBar/Statuses.tsx [hunk 6]
```diff
@@ -10,27 +10,28 @@ import { ParcelsTableRow } from "@/app/parcels/getParcelsTableData";
...
+    "Out for Delivery",
     "Delivered",
     "Delivery Failed",
     "Delivery Cancelled",
     "Fulfilled with Trussell Trust",
-    "Shipping Labels Downloaded",
-    "Shopping List Downloaded",
-    "Out for Delivery",
     "Request Deleted",
-] as const;
```
[the remark this answers is not in the document]
Comment (AmeliaIsCoding) [reply src/app/parcels/ActionBar/Statuses.tsx:31]: In the DB, status.eventName is a string, not an enum of these statuses, so I think that in order to sort the events by eventNames, we need to compare to an ordered list of strings, not an ordered list of enums. Could change the DB type but that seems scary?

[the remark this answers is not in the document]
Comment (rickaw) [reply src/app/parcels/ActionBar/Statuses.tsx:31]: seems like there isn't a _good_ solution for this in typescript - happy to leave it!

**Review by rickaw — APPROVED** [review]

**Pull request merged**